aioxmpp/pep/service.py

```diff
@@ -0,0 +1,363 @@
+            receive notification without explicit subscription.
+        :raises RuntimeError: if a handler for `node_namespace` is already
+            set.
+        :returns: a :class:`RegisteredPEPNode` instance representing
+            the claim.
+
+        This registers `node_namespace` as feature for service discovery
+        unless ``register_feature=False`` is passed.
+        """
+        if node_namespace in self._pep_node_claims:
+            raise RuntimeError(
+                "claiming already claimed node"
+            )
+        registered_node = RegisteredPEPNode(
+            self,
+            node_namespace,
+            register_feature=register_feature,
+            notify=notify,
+        )
+
+        finalizer = weakref.finalize(
+            registered_node,
+            weakref.WeakMethod(registered_node._unregister)
+        )
+        # we cannot guarantee that disco is not cleared up already
+        finalizer.atexit = False
+        self._pep_node_claims[node_namespace] = registered_node
+
+        return registered_node
+
+    def _unclaim(self, node_namespace):
+        self._pep_node_claims.pop(node_namespace)
+
+    @asyncio.coroutine
+    def _check_for_pep(self):
+        # XXX: should this be done when the stream connects
+        # and we use the cached result later on (i.e. disable
+        # the PEP service if the server does not support PEP)
+        disco_info = yield from self._disco_client.query_info(
+            self.client.local_jid.bare()
+        )
+
+        for item in disco_info.identities.filter(attrs={"category": "pubsub"}):
+            if item.type_ == "pep":
+                break
+        else:
+            raise RuntimeError("server does not support PEP")
+
+    @service.depsignal(aioxmpp.PubSubClient, "on_item_published")
+    def _handle_pubsub_publish(self, jid, node, item, *, message=None):
+        try:
+            registered_node = self._pep_node_claims[node]
+        except KeyError:
+            return
+
+        # TODO: handle empty payloads due to (mis-)configuration of
+        # the node specially.
+        registered_node.on_item_publish(jid, node, item, message=message)
+
+    def publish(self, node, data, *, id_=None):
+        """
+        Publish an item `data` in the PubSub node `node` on the
+        PEP service associated with the user's JID.
+
+        :param node: The PubSub node to publish to.
+        :param data: The item to publish.
+        :type data: An XSO representing the paylaod.
+        :param id_: The id the published item shall have.
+        :returns: The PubSub id of the published item or
+            :data:`None` if it is unknown.
+
+        If no `id_` is given it is generated by the server (and may be
+        returned).
+        """
+        yield from self._check_for_pep()
+        return (yield from self._pubsub.publish(None, node, data, id_=id_))
+
+
+class RegisteredPEPNode:
```

Comment:
The tests are relatively extensive already and should get most cases. I feel uncomfortable about the complexity anyway (interaction of the close method and the weakref cleanup). Especially, because disco_server.{un,}register_feature raise exceptions if they are called twice.